compiler/ballerina-lang/src/main/java/io/ballerina/projects/JBallerinaBackend.java

```diff
@@ -364,37 +543,168 @@ public void performCodeGen(ModuleContext moduleContext, CompilationCache compila
+            throw new IllegalStateException("Missing generated jar, module: " + moduleContext.moduleName());
+        }
+        String jarFileName = getJarFileName(moduleContext) + JAR_FILE_NAME_SUFFIX;
+        if (bLangPackage.symbol.shouldGenerateDuplicateBIR) {
+            jarFileName = getJarFileName(moduleContext) + "_OPTIMIZED" + JAR_FILE_NAME_SUFFIX;
+        }
+        try {
+            ByteArrayOutputStream byteStream = JarWriter.write(compiledJarFile, getResources(moduleContext));
+            if (!this.packageContext.project().buildOptions().skipTests()) {
+                moduleContext.getCompilationCache().cachePlatformSpecificLibrary(this, jarFileName, byteStream, true);
+            } else {
+                optimizedJarStreams.putIfAbsent(jarFileName, byteStream);
+            }
+        } catch (IOException e) {
+            throw new ProjectException("Failed to cache generated jar, module: " + moduleContext.moduleName());
+        }
+
+        // TODO merge this with performCodeGen
+        // skip generation of the test jar if --with-tests option is not provided
+        if (moduleContext.project().buildOptions().skipTests() || !bLangPackage.hasTestablePackage()) {
             return;
         }
 
+        // Both invocation data are the same
+        // TODO do the merging inside usedBIRNodeAnalyzer
+        bLangPackage.testablePkgs.get(0).symbol.invocationData = bLangPackage.symbol.invocationData;
+
         String testJarFileName = jarFileName + TEST_JAR_FILE_NAME_SUFFIX;
         CompiledJarFile compiledTestJarFile = jvmCodeGenerator.generateTestModule(bLangPackage.testablePkgs.get(0),
                 isRemoteMgtEnabled);
         try {
             ByteArrayOutputStream byteStream = JarWriter.write(compiledTestJarFile, getAllResources(moduleContext));
-            compilationCache.cachePlatformSpecificLibrary(this, testJarFileName, byteStream);
+            moduleContext.getCompilationCache().cachePlatformSpecificLibrary(this, testJarFileName, byteStream, true);
         } catch (IOException e) {
             throw new ProjectException("Failed to cache generated test jar, module: " + moduleContext.moduleName());
         }
     }
 
+    private void duplicateCodegen(ModuleContext moduleContext, boolean isRemoteMgtEnabled) {
+        BLangPackage bLangPackage = moduleContext.bLangPackage();
+        BIRNode.BIRPackage optimizableBirPkg = bLangPackage.symbol.bir;
+        bLangPackage.symbol.bir = bLangPackage.symbol.duplicateBir;
+        interopValidator.validate(moduleContext.moduleId(), this, bLangPackage);
+        if (bLangPackage.getErrorCount() > 0) {
+            return;
+        }
+        CompiledJarFile originalJarFile = jvmCodeGenerator.generate(bLangPackage, false, isRemoteMgtEnabled);
+        bLangPackage.symbol.bir = optimizableBirPkg;
+
+        if (originalJarFile == null) {
+            throw new IllegalStateException("Missing generated jar, module: " + moduleContext.moduleName());
+        }
```

Comment:
Shall we move this null check up so that we don't change the state of bLangPackage if we have an invalid state?
```suggestion
        if (originalJarFile == null) {
            throw new IllegalStateException("Missing generated jar, module: " + moduleContext.moduleName());
        }

        bLangPackage.symbol.bir = optimizableBirPkg;
```

Reply:
Is this to allow the compilation to continue, even if an `IllegalStateException` was to be thrown?

If so the best way to handle it will be to swap the BIRs. Since we are generating 2 duplicate BIR pkgs, only one of them will be referred by other the BIR pkgs. `optimizableBirPkg` is the one that is being referred by other pkgs. During duplicate codegen, we swap the referred and duplicate BIRs. And swap them back again for optimized codegen. Therefore, if we want to continue the process even after a failed codegen, it is better to swap them again.

Reply:
After a discussion with @heshanpadmasiri, we decided to move the mutation of the `bLangPackage.symbol.bir` after the exception throw logic to preserve the erroneous state. This way it is easier to debug the code in the future and identify the erroneous parts. Change was done in eb164e9b0bfd3dac61e1b00a86d30e96f11b33a2.